cr-api/cr/api/server.py

```diff
@@ -1,58 +1,395 @@
-import cherrypy
+from collections import defaultdict
+from functools import wraps
+import hashlib
+import itertools
 import json
+import math
 import sys
+
+import cherrypy
+from cherrypy import expose
+from cherrypy import tools
+import numpy as np
+
 from cr.db.store import global_settings as settings, connect
 
-class Root(object):
+
+SESSION_KEY = 'id'
+
+
+class User(dict):
+
+    def serialize(self):
+        return {
+            'email': self['email'],
+            'first_name': self['first_name'],
+            'last_name': self['last_name'],
+            'latitude': self['latitude'],
+            'longitude': self['longitude'],
+        }
+
+    @staticmethod
+    def deserialize(data):
+        return {
+            'email': data['email'],
+            'first_name': data['first_name'],
+            'last_name': data['last_name'],
+            'latitude': data['latitude'],
+            'longitude': data['longitude'],
+        }
+
+
+def calculate_distance_miles(p1, p2):
+    # Based on public domain code from:
+    #    https://www.johndcook.com/blog/python_longitude_latitude/
+
+    degrees_to_radians = math.pi/180.0
+    (lat1, long1), (lat2, long2) = p1, p2
+
+    # phi = 90 - latitude
+    phi1 = (90.0 - lat1)*degrees_to_radians
+    phi2 = (90.0 - lat2)*degrees_to_radians
+
+    # theta = longitude
+    theta1 = long1*degrees_to_radians
+    theta2 = long2*degrees_to_radians
+
+    cos = (np.sin(phi1)*np.sin(phi2)*np.cos(theta1 - theta2) +
+           np.cos(phi1)*np.cos(phi2))
+    arc = np.arccos(cos)
+
+    # multiply arc by:
+    #    3960 for miles
+    #    6373 for kilometers
+    return arc * 3960
+
+
+distances_cache = {}
+
+
+class Geo(object):
+
+    def __init__(self, users):
+        self.distances = []
+        points = [self._point(user) for user in users]
+        pairs = list(itertools.combinations(points, 2))
+        for p1, p2 in pairs:
+            self.distances.append(self._get_distance(p1, p2))
+
+    def _point(self, user):
+        return (int(float(user['latitude'])), int(float(user['longitude'])))
+
+    def _get_distance(self, p1, p2):
+        # Try to use the already cached distance because:
+        #    distance(p1, p2) == distance(p2, p1)
+        if (p1, p2) in distances_cache:
+            return distances_cache[(p1, p2)]
+        if (p2, p1) in distances_cache:
+            return distances_cache[(p2, p1)]
+        try:
+            distance = int(calculate_distance_miles(p1, p2))
+        except ValueError:
+            distance = None
+        distances_cache[(p1, p2)] = distance
+        distances_cache[(p2, p1)] = distance
+        return distance
+
+
+class GeoDetails(Geo):
```

Comment:
Why subclass `Geo`?

Reply:
copy/pasta ;)

I noticed & fixed that yesterday.

- https://github.com/dianaclarke/crunch-fitness/commit/15b4f57fd799288d94356cadc49b1112b8dd84d1

Reply:
Ooops, I see what you mean now. I thought you were talking about my `StatsMixIn` typo from yesterday.

I subclassed Geo to reuse the `_point ` and `_get_distance ` methods, but I really should have cleaned up those two classes more. Made their responsibilities clearer, decoupled them, etc.

That code was written quickly when I realized last minute that I misunderstood the assignment. Excuses, excuses ;)

Reply:
Ah, I missed that. Fair enough.

I'm a recovering object zealot, and would have made those free-standing functions.

Reply:
PS. What I would do differently:

1. Scrub the data on initial import (and subsequent inserts) so that `_point` wasn't needed at all. That is, store lat/log as floats (or ints) in monngo rather than the current strings.

2. Perhaps use `beaker`/`dogpile.cache` & `memcached` to cache the distances, and then `_get_distance` would go away.  Something like:

```
@cache.cache('distances')
def get_distance(p1, p2):
    return int(calculate_distance_miles(p1, p2))
```

And then `GeoDetails` would no longer inherit from `Geo`.